### ISSUE TITLE
Auto-DNS not working properly

### ISSUE THREAD

**Issue by rslid** [cytopia/devilbox#132]
Hi,

I have pulled today latest image and all works well except auto-dns feature. Under php 7.1, apache 2.4, Docker version 17.09.0-ce, build afdb6d4 and osx high sierra when I set `nameserver 127.0.0.1` in resolv.conf, flush the dns cache, delete the container and run compose up virtualhosts are not properly resolved. If I place them in hosts file all works well. Any suggestions for further debugging? I left DEBUG_COMPOSE_ENTRYPOINT=1 and checked the log but everything seems fine, no errors on the screen so far. 

Also after wifi reconnect on osx of dns are changed and file completely wiped, so this means that I have to add localhost on each wifi reconnect...

**Comment by cytopia** [comment]
@rslid 

1. Did you change the Nameserver port to 53 in `.env`?
2. Can you try again with current PR branch: `REL-0.12`

Additional instructions for how to use new features in current PR can be found here: https://github.com/cytopia/devilbox/pull/130

**Comment by SethTompkins** [comment]
This is also failing for me: macOS High Sierra 10.13

Bind is running and sets up my wildcard domain correctly, however all requests to the site timeout. Interestingly, if I use dnsmasq to forward requests for *.test to 127.0.0.1 then the site works.

**Comment by cytopia** [comment]
@SethTompkins Did you change Bind port to `53` in `.env` file?

```shell
HOST_PORT_BIND=53
```

**Comment by SethTompkins** [comment]
@cytopia I do have that set, here is my `/etc/resolv.conf`

```
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
search openvpn
nameserver 127.0.0.1
nameserver 8.8.8.8
```

**Comment by cytopia** [comment]
@SethTompkins looks good so far. 

1. What about `HOST_PORT_BIND`?
2. Test local nameserver

Can you do the following (Replace `loc` to whatever you have set `TLD_SUFFIX`) and add the output here.

```shell
dig foo.loc @127.0.0.1
```

**Comment by SethTompkins** [comment]
@cytopia, `HOST_PORT_BIND` is set to 53, here is the output for that command:

```
$ dig wax-app.test @127.0.0.1

; <<>> DiG 9.9.7-P3 <<>> wax-app.test @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 53029
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;wax-app.test.			IN	A

;; ANSWER SECTION:
wax-app.test.		604800	IN	A	172.16.238.11

;; AUTHORITY SECTION:
test.			604800	IN	NS	.

;; Query time: 2 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Thu Oct 19 09:12:46 EDT 2017
;; MSG SIZE  rcvd: 70
```

When I try to hit that site in my browser though, it spins until I eventually get a `ERR_CONNECTION_TIMED_OUT`

**Comment by SethTompkins** [comment]
Additionally, in the devilbox admin it displays this message for the virtual host:
```
No Host DNS record found. Add the following to /etc/hosts:
127.0.0.1 wax-app.test
```

**Comment by cytopia** [comment]
@SethTompkins everything looks fine.

Can you make sure, that you are on the latest `master` branch, as well as have the latest Docker images.

Please do the commands in the described order and in case of differences, paste those please.

```shell
git checkout master
git fetch --all
git diff origin/master > /tmp/paste-here.txt
git pull origin master

./update-docker.sh
```

If there were changes, try to restart the devilbox:

```
docker-compose stop
docker-compose kill
docker-compose rm -f
docker-compose up
```

**Comment by cytopia** [comment]
Btw, can you also check if your local installation meets the following requirements:

https://github.com/cytopia/devilbox/blob/master/docs/README.md#4-requirements

**Comment by cytopia** [comment]
Additionally, if it still does not work:

1. Use Chrome
2. Use an incognito window
3. Open Developer console
4. Disable Cache

Are there any error messages in the developer console from chrome?

**Comment by SethTompkins** [comment]
@cytopia I ran all of the commands specified, there were no changes. The update script ran for a while but after running it and the docker-compose commands I am still in the same situation.

Using an Incognito window in Chrome w/ dev tools and cache disabled yields the same results as well - no error messages in the console.

I am on latest macOS with latest version of Docker for Mac.

**Comment by SethTompkins** [comment]
Only other piece of info that might be helpful: if I use dnsmasq via homebrew to forward all *.test requests to 127.0.0.1 the site works.

**Comment by cytopia** [comment]
I don't quiet understand how the devilbox DNS can listen on `:53` as well as the dnsmasq (which should also listen on `:53`).

Is dnsmasq part of MacOS nowadays?

**Comment by cytopia** [comment]
Btw, can you actually reach `172.16.238.11`. Try to open it in a browser `http://172.16.238.11`

**Comment by SethTompkins** [comment]
I cannot reach `172.16.238.11`, it times out just like trying to reach one of the sites.

dnsmasq is not part of macOS, I installed it via homebrew. I am going to try to completely remove it and see if that fixes my issues.

It has been stopped all of the times when I have been testing this though - possibly some sort of configuration that is causing a collision.

**Comment by cytopia** [comment]
@SethTompkins not required to check anymore.

The problem is that `172.16.238.11` is not reachable. This is the HTTPD docker Container (Nginx or Apache).

So it must be some problem with Docker bridged network on OSX. Can you ping/reach any of the following IP addresses (when the devilbox is running):
```
        ipv4_address: 172.16.238.100
        ipv4_address: 172.16.238.10
        ipv4_address: 172.16.238.11
        ipv4_address: 172.16.238.12
        ipv4_address: 172.16.238.13
        ipv4_address: 172.16.238.14
        ipv4_address: 172.16.238.15
        ipv4_address: 172.16.238.16
```

**Comment by cytopia** [comment]
I just found the documentation in the Docker docs:

https://docs.docker.com/docker-for-mac/networking/#there-is-no-docker0-bridge-on-macos

At least the problem is finally identified

**Comment by SethTompkins** [comment]
What is strange then is that I can connect to the mysql instance from my host computer via SequelPro, shouldn't this issue prevent that as well?

**Comment by cytopia** [comment]
Yes, you can connect to it via `127.0.0.1` or?, but not via `172.16.238.12:3306` (which is the IP address of the MySQL container).

So all ports forwarded to your host OS work, but connecting directly to the Docker container's IP addresses does not seem to work.

**Comment by SethTompkins** [comment]
Correct, I am connecting with the host as `127.0.0.1`, which makes sense why DNS works with dnsmasq as it is set to simply forward all requests on my wildcard tld to `127.0.0.1`

**Comment by cytopia** [comment]
Ok, I will try to find a solution, so that DNS is forwarding requests to `127.0.0.1` instead of the docker IP

**Comment by cytopia** [comment]
Can you give this a try: https://github.com/cytopia/devilbox/pull/138

Branch: `DVL-013-fix-dns`

**Comment by SethTompkins** [comment]
Works like a charm for me :)

Thanks for all of your help and quick replies!

**Comment by cytopia** [comment]
Good to hear. I will keep the PR open, until its been also proven on Linux and Windows.

**Comment by rslid** [comment]
Guys sorry for the late join in the discussion and thanks both of you for resolving this issue. I can confirm that DVL-013-fix-dns works for me perfectly.

**Comment by SethTompkins** [comment]
@cytopia update on this: after having to restart my devilbox instance the DNS no longer seems to work. The devilbox admin reports that there are no records for the sites as well. Bind is reported as running in the admin though.

**Comment by SethTompkins** [comment]
To add to that last note, now I am getting an `ERR_NAME_NOT_RESOLVED` rather than a timeout.

**Comment by cytopia** [comment]
@SethTompkins don't really understand what you mean. On the provided screenshot everything looks good.

Apart from that, did you do a full restart:
```shell
docker-compose stop
docker-compose kill
docker-compose rm -f
docker-compose up
```

**Comment by SethTompkins** [comment]
@cytopia yup, just did all those commands again for good measure, DNS still not registering. Only thing I notice form that screen is that bind does not have `127.0.0.1` listed after its internal ip.

**Comment by cytopia** [comment]
Then were do you get those errors? The screenshot looks all good

**Comment by SethTompkins** [comment]
Screengrab from the vhosts section:
![image](https://user-images.githubusercontent.com/3925226/31841614-398c8a62-b5b8-11e7-89eb-f05619795956.png)

If I try to access one of those sites I do not get a timeout like before, it simply tells me "(project name's) server DNS address could not be found."

**Comment by cytopia** [comment]
Was your `/etc/resolv.conf` auto-updated by OSX or do you still have the devilbox dns server present?

**Comment by SethTompkins** [comment]
`nameserver 127.0.0.1` is in that file, here is the output from `scutil --dns`

```
resolver #1
  search domain[0] : spiiider.com
  nameserver[0] : 127.0.0.1
  nameserver[1] : 8.8.8.8
  if_index : 5 (en4)
  flags    : Scoped, Request A records, Request AAAA records
  reach    : 0x00030002 (Reachable,Local Address,Directly Reachable Address)

resolver #2
  search domain[0] : spiiider.com
  nameserver[0] : 127.0.0.1
  nameserver[1] : 8.8.8.8
  nameserver[2] : 8.8.4.4
  if_index : 6 (en0)
  flags    : Scoped, Request A records, Request AAAA records
  reach    : 0x00030002 (Reachable,Local Address,Directly Reachable Address)
```

**Comment by cytopia** [comment]
What does dig say? `dig wax-app.test @127.0.0.1`

**Comment by SethTompkins** [comment]
Dig output:

```
; <<>> DiG 9.9.7-P3 <<>> wax-app.test @127.0.0.1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 16597
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;wax-app.test.			IN	A

;; ANSWER SECTION:
wax-app.test.		604800	IN	A	127.0.0.1

;; AUTHORITY SECTION:
test.			604800	IN	NS	.

;; Query time: 1 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Oct 20 17:07:04 EDT 2017
;; MSG SIZE  rcvd: 70
```

**Comment by SethTompkins** [comment]
@cytopia here is the output of `docker-compose logs` for bind:

```
Attaching to devilbox_bind_1
bind_1   | [INFO] Adding wildcard DNS record: '*.test' -> '127.0.0.1'
bind_1   | root $ echo 'include "/etc/bind/custom-named.conf.test";' >> /etc/bind/named.conf
bind_1   | [INFO] Adding custom DNS forwarder: 8.8.8.8,8.8.4.4
bind_1   | root $ sed -i'' 's/^};$/	forwarders {
bind_1   | 		8.8.8.8;
bind_1   | 		8.8.4.4;
bind_1   | 	};
bind_1   | };/' /etc/bind/named.conf.options
bind_1   | [INFO] Starting BIND 9.10.3
```

**Comment by cytopia** [comment]
@SethTompkins something must have changed on your side in between. It did work at some point or?

@rslid can you please check again if it still works for your.

**Comment by SethTompkins** [comment]
@cytopia I seem to have tracked down the issue:

Occasionally at work I have to use a dumpster fire of a VPN program - Forticlient. When that program is active it messes up my local DNS and devilbox bind does not work. Sorry for wasting your time with this, looks like the issue is entirely on my end.

**Comment by rslid** [comment]
@cytopia I have tested now on completely new osx as I don't have access to old setup and branch DVL-013-fix-dns seems to not work for me. I get the same results as @SethTompkins mentioned above, I checked resolver.conf, stopped/killed and up'ed all containers and etc but without any success. I don't have any vpn's or other software which could be messing up this. I used dig and response was the same as above mentioned ones.

I had the issue with httpd container where MAIN_VHOST_DISABLE was set to empty string when it should be 0 or 1 and container exited, I fixed it by explicitly setting DEVILBOX_UI_DISABLE in .env and this was resolved. However I cant figure out for now whats wrong when all logs show proper setup/response.

**Comment by cytopia** [comment]
@SethTompkins  @rslid I will do some more digging into this topic once I have finished working on the new PHP-FPM images.

**Comment by kazuo** [comment]
I encountered the same issue and tried the DVL-013-fix-dns branch (with merging master). The branch works for me with a few caveats. By the way, I'm running macOS 10.13.2 (High Sierra) with Docker Edge. 

When using the Auto-DNS configuration, I can't use `local` as a search domain with any `docker-compose` command (i.e. `up`, `down`, `rm`, etc). I either need to remove the `local` search domain or rename it to something from `local` to anything else... like `localhost` then the commands would continue to go through. (At first, I thought it was coincidence, but running any `docker-compose` command in devilbox it looked like the command would hang until I elimited the `local` search domain.)

That said, once devilbox was up, in order for Auto-DNS to work, I had to re-add the `local` search domain. Once I did that, it resolved just fine (and my webpage loaded without issue).

Also, I didn't realize this in macOS, but modifying `/etc/resolv.conf` doesn't work. I had to modify my DNS through the normal DNS config under Network. Once you hit `Apply`, `/etc/resolv.conf` updates immediately. So all my testing I did was through that. 

By the way, this branch with the caveats still work even if I restart docker or bring devilbox `down` and `up` again.

**Comment by cytopia** [comment]
@kazuo good to know and thanks for reporting.

**Comment by cytopia** [comment]
@kazuo @rslid @SethTompkins 
I have rebuild the Bind image. Could you give it a try (described here: https://github.com/cytopia/devilbox/issues/163)

**Comment by kazuo** [comment]
Pointed `bind` to `release-0.12`, ran `./update-docker.sh`, and before doing so ensured that `127.0.0.1` and `local` was in DNS servers and Search Domains respectively.

* `docker-compose up` worked as expected
* Going to my website using Auto-DNS worked as expected

Working great so far!

**Comment by rslid** [comment]
@cytopia 
I have tested this on my main machine where initial issues with auto-dns feature appeared:
1. Update docker-compose.yml to point to bind release-0.12 (I also used DVL-013-fix-dns branch) 
2. Stopped/removed all containers and brought them up again
3. Created new entry in data/www and cleaned hosts file from old records

I'm happy to report that now everything works perfectly, bravo!

**Comment by cytopia** [comment]
Pr has been merged.

**Comment by 06tech** [comment]
Sorry to re-visit this issue but I am getting auto DNS resolve issues on my Mac OS X (Monterey) 12.5.1... I'm running Devilbox  v2.2.0 (Apr 2022). Image for BIND is cytopia/bind:alpine-0.28 

When I restart devilbox with a stop/rm -f/up DNS will start to fail when resolving the .loc hostnames with ping responding with ' 'Unknown host'.

The only way to fix it is to remove/re-add 127.0.0.1 in the Networks/DNS settings. Is there something else I need to do to avoid having to do this?

Thanks!

**Comment by stollz** [comment]
I kike this thread as a good source of information how to tackle down auto-dns issues. But I am facing the same issue as @06tech : I need to remove/re-add 127.0.0.1 in the Macos Network DNS settings and everything is fine. Any idea how to fix this? May be a command line alternative ?
Thanks a lot for any help!
